runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java

```diff
@@ -234,6 +239,97 @@ public void testOnTimerCalled() {
                 TimeDomain.EVENT_TIME)));
   }
 
+  @Test
+  public void testBackwardsInTimeNoSkew() {
+    SkewingDoFn fn = new SkewingDoFn(Duration.ZERO);
+    DoFnRunner<Duration, Duration> runner =
+        new SimpleDoFnRunner<>(
+            null,
+            fn,
+            NullSideInputReader.empty(),
+            new ListOutputManager(),
+            new TupleTag<Duration>(),
+            Collections.<TupleTag<?>>emptyList(),
+            mockStepContext,
+            null,
+            WindowingStrategy.of(new GlobalWindows()));
+
+    runner.startBundle();
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.ZERO, new Instant(0)));
+    thrown.expect(UserCodeException.class);
+    thrown.expectCause(isA(IllegalArgumentException.class));
+    thrown.expectMessage("must be no earlier");
+    thrown.expectMessage(
+        String.format("timestamp of the current input (%s)", new Instant(0).toString()));
+    thrown.expectMessage(
+        String.format(
+            "the allowed skew (%s)", PeriodFormat.getDefault().print(Duration.ZERO.toPeriod())));
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.millis(1L), new Instant(0)));
+  }
+
+  @Test
+  public void testSkew() {
+    SkewingDoFn fn = new SkewingDoFn(Duration.standardMinutes(10L));
+    DoFnRunner<Duration, Duration> runner =
+        new SimpleDoFnRunner<>(
+            null,
+            fn,
+            NullSideInputReader.empty(),
+            new ListOutputManager(),
+            new TupleTag<Duration>(),
+            Collections.<TupleTag<?>>emptyList(),
+            mockStepContext,
+            null,
+            WindowingStrategy.of(new GlobalWindows()));
+
+    runner.startBundle();
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.millis(1L), new Instant(0)));
+    thrown.expect(UserCodeException.class);
+    thrown.expectCause(isA(IllegalArgumentException.class));
+    thrown.expectMessage("must be no earlier");
+    thrown.expectMessage(
+        String.format("timestamp of the current input (%s)", new Instant(0).toString()));
+    thrown.expectMessage(
+        String.format(
+            "the allowed skew (%s)",
+            PeriodFormat.getDefault().print(Duration.standardMinutes(10L).toPeriod())));
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.standardHours(1L), new Instant(0)));
+  }
+
+  @Test
+  public void testInfiniteSkew() {
+    SkewingDoFn fn = new SkewingDoFn(Duration.millis(Long.MAX_VALUE));
```

Comment:
Is there a test for the case where actual skew < allowed skew and things should pass?